### PR TITLE
llmproxy: Add sanitization of request body and extend logging

### PR DESCRIPTION
**Pull request by eseliger**
This PR adds parsing of the request and response from/to Anthropic.
- By that, we can make sure to only forward fields that we know of, and prevent usage of future fields without us gating them.
- We can also gate model access to specific ones in a follow up PR
- We can start applying per-model rate limits after this PR, technically
- We get more logging, including character counts
- It prevents users setting the metadata field, which is meant for internal use to identify traffic

Since things got quite messy in the main package, I also moved the handler into a separate package as it grew too much in size to stay comfortably in main.

Screenshot 2023-05-15 at 21 36 35@2x

Closes https://github.com/sourcegraph/sourcegraph/issues/51362

## Test plan

Verified things still work correctly E2E locally, and adjusted some tests for the Anthropic decoder.

### CODE REVIEW

File: enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+type anthropicResponse struct {
+	Completion string `json:"completion"`
+	StopReason string `json:"stop_reason"`
+}
+
+type anthropicCompletionEvent struct {
+	Completion string `json:"completion"`
+	StopReason string `json:"stop_reason"`
+	LogID      string `json:"log_id"`
+}
```
[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go:184]: yep took all of that from here: https://console.anthropic.com/docs/api/reference

**Review by bobheadxi — APPROVED** [review]

File: enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		var responseBuf bytes.Buffer
+		respBody := io.TeeReader(resp.Body, &responseBuf)
+		// Forward response to client.
+		_, _ = io.Copy(w, respBody)
+
+		// Now try to parse the request we saw, if it was non-streaming, we can simply parse
+		// it as JSON.
+		if !body.Stream {
+			var res anthropicResponse
+			if err := json.NewDecoder(&responseBuf).Decode(&res); err != nil {
```
Comment (keegancsmith) [inline enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go:127]: very minor: you already have a bytes.Buffer, using NewDecoder will just setup another one. Better to use json.Unmarshal(responseBuf.Bytes())

File: enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+				logger.Error("failed to parse anthropic response as JSON", log.Error(err))
```
Comment (keegancsmith) [inline enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go:128]: are you monitoring the logs of llm-proxy? For example I imagine this line will be noisy due to incomplete responseBufs thanks to the user cancelling the request.

Comment (eseliger) [reply enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go:128]: we should be :D still figuring out operations. Will mute this if it turns out to be noisy :)

File: enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+		// Forward status code.
+		upstreamStatusCode = resp.StatusCode
+		w.WriteHeader(resp.StatusCode)
+
+		// Set up a buffer to capture the response as it's streamed and sent to the client.
+		var responseBuf bytes.Buffer
+		respBody := io.TeeReader(resp.Body, &responseBuf)
+		// Forward response to client.
+		_, _ = io.Copy(w, respBody)
```
[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go:121]: thanks, will follow up with fixes for these two comments!

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go:121]: Hmm all the solutions to this I produced so far didn't really make me super confident in the correctness of them.. 😁 so easy to create a leaky goroutine or a memory leak 😢

[the remark this answers is not in the document]
Comment (keegancsmith) [reply enterprise/cmd/llm-proxy/internal/httpapi/anthropic.go:121]: yeah I reckon stick to just copying to a buffer then

**Pull request merged**